firebase-vscode/common/messaging/broker.ts

```diff
@@ -0,0 +1,86 @@
+import { Listener, Message, MessageListeners } from "./types";
+import { Webview } from "vscode";
+
+const isObject = (val: any): boolean => typeof val === "object" && val !== null;
+
+type Receiver = {} | Webview;
+
+export abstract class Broker {
+  protected readonly listeners: MessageListeners = {};
+
+  abstract sendMessage(message: string, data: any[]): void;
+  registerReceiver(receiver: any): void { }
+
+  addListener(message: string, cb: Listener): void {
+    if (!this.listeners[message]) {
+      this.listeners[message] = { listeners: [] };
+    }
+    this.listeners[message].listeners.push(cb);
+  }
+
+  executeListeners(data: Message) {
```

Comment:
Naming is a bit hard to follow here, since `data` is a message and message has a field `data`

Reply:
`Message` also has a field `message`, which doesn't help. I renamed Message.message to Message.command and this to `message: Message`.